File: infrastructures/ddd/python/src/infrastructure/queues/in_memory_event_bus.py
"""
In-memory event bus implementation for development and testing.
In production, replace with actual message queue (RabbitMQ, Kafka, AWS SQS, etc.).
"""

import asyncio
from typing import Any, Callable, Dict, List
from infrastructure.queues.infra import IEventPublisher, IEventSubscriber


class InMemoryEventBus(IEventPublisher, IEventSubscriber):
    """
    In-memory event bus for development/testing.
    
    This implementation:
    - Stores subscribers in memory
    - Publishes events synchronously to all subscribers
    - Suitable for single-process applications
    
    For production, use actual message queue:
    - RabbitMQ with aio-pika
    - Kafka with aiokafka
    - AWS SQS/SNS with aioboto3
    - Azure Service Bus with azure-servicebus
    - Redis Pub/Sub with aioredis
    
    Example with RabbitMQ (aio-pika):
    ```python
    import aio_pika
    import json
    
    class RabbitMQEventBus(IEventPublisher, IEventSubscriber):
        def __init__(self, p_connection_string: str):
            self._connection_string = p_connection_string
            self._connection = None
            self._channel = None
        
        async def connect_async(self):
            self._connection = await aio_pika.connect_robust(self._connection_string)
            self._channel = await self._connection.channel()
        
        async def publish_async(self, p_event: Any, p_topic: str) -> None:
            exchange = await self._channel.declare_exchange(
                p_topic,
                aio_pika.ExchangeType.FANOUT
            )
            
            message = aio_pika.Message(
                body=json.dumps(p_event.__dict__).encode()
            )
            
            await exchange.publish(message, routing_key="")
    ```
    """
    
    def __init__(self):
        """Initialize the in-memory event bus."""
        self._subscribers: Dict[str, List[Callable]] = {}
    
    async def publish_async(self, p_event: Any, p_topic: str) -> None:
        """
        Publishes an event to all subscribers of the topic.
        
        Args:
            p_event: The event to publish
            p_topic: The topic to publish to
        """
        handlers = self._subscribers.get(p_topic, [])
        
        # Call all handlers asynchronously
        tasks = [handler(p_event) for handler in handlers]
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)
    
    async def subscribe_async(
        self,
        p_topic: str,
        p_handler: Callable[[Any], None]
    ) -> None:
        """
        Subscribes a handler to a topic.
        
        Args:
            p_topic: The topic to subscribe to
            p_handler: The handler function for events
        """
        if p_topic not in self._subscribers:
            self._subscribers[p_topic] = []
        
        self._subscribers[p_topic].append(p_handler)

File: infrastructures/ddd/typescript/src/infrastructure/queues/InMemoryEventBus.ts
import { IEventPublisher } from './infra/IEventPublisher';
import { IEventSubscriber } from './infra/IEventSubscriber';

/**
 * In-memory event bus implementation for demonstration purposes.
 * Queue implementations:
 * - Live in /infrastructure/queues/
 * - Are concrete implementations (no /impl/ subdirectory)
 * - Implement IEventPublisher and IEventSubscriber
 * 
 * In production, replace this with:
 * - RabbitMQ
 * - Azure Service Bus
 * - AWS SQS/SNS
 * - Google Cloud Pub/Sub
 * - Apache Kafka
 * 
 * This implementation stores handlers in memory and invokes them immediately.
 * Real implementations would:
 * - Serialize events to queue
 * - Handle retries and dead-letter queues
 * - Provide durability guarantees
 * - Support distributed systems
 */
export class InMemoryEventBus implements IEventPublisher, IEventSubscriber {
  private readonly _handlers: Map<string, Array<(event: any) => Promise<void>>>;

  constructor() {
    this._handlers = new Map();
  }

  /**
   * Publishes a domain event to the specified topic.
   * In this in-memory implementation, immediately invokes all subscribed handlers.
   * In production, would send to message queue.
   */
  public async PublishAsync<TEvent>(p_event: TEvent, p_topic: string): Promise<void> {
    console.log(`[InMemoryEventBus] Publishing event to topic '${p_topic}':`, p_event);

    const handlers = this._handlers.get(p_topic) || [];
    
    // Invoke all handlers for this topic
    for (const handler of handlers) {
      try {
        await handler(p_event);
      } catch (error) {
        console.error(`[InMemoryEventBus] Handler failed for topic '${p_topic}':`, error);
        // In production, implement retry logic and dead-letter queue
      }
    }
  }

  /**
   * Subscribes to a topic and handles events with the provided handler.
   * In this in-memory implementation, stores handler in memory.
   * In production, would subscribe to message queue.
   */
  public async SubscribeAsync<TEvent>(
    p_topic: string,
    p_handler: (p_event: TEvent) => Promise<void>
  ): Promise<void> {
    console.log(`[InMemoryEventBus] Subscribing to topic '${p_topic}'`);

    if (!this._handlers.has(p_topic)) {
      this._handlers.set(p_topic, []);
    }

    this._handlers.get(p_topic)!.push(p_handler);
  }

  /**
   * Unsubscribes all handlers for a topic.
   * Useful for testing and cleanup.
   */
  public UnsubscribeAll(p_topic: string): void {
    console.log(`[InMemoryEventBus] Unsubscribing all handlers from topic '${p_topic}'`);
    this._handlers.delete(p_topic);
  }

  /**
   * Gets the number of subscribers for a topic.
   * Useful for testing and diagnostics.
   */
  public GetSubscriberCount(p_topic: string): number {
    return (this._handlers.get(p_topic) || []).length;
  }
}
